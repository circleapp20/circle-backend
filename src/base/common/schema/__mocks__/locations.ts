import { BaseModel } from './baseModel';
import { Users } from './users';

export class Locations extends BaseModel {
	name: string;
	latitude: number;
	longitude: number;
	address: string;
	isVerified: boolean;
	place: Locations;
	places: Locations[];
	user: Users;
}
