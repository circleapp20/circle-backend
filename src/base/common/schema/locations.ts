import * as typeorm from 'typeorm';
import { BaseModel } from './baseModel';
import { Users } from './users';

@typeorm.Entity()
export class Locations extends BaseModel {
	@typeorm.Column()
	name: string;

	@typeorm.Column('float', { default: 0 })
	latitude: number;

	@typeorm.Column('float', { default: 0 })
	longitude: number;

	@typeorm.Column()
	address: string;

	@typeorm.Column('boolean', { default: false })
	isVerified: boolean;

	@typeorm.ManyToOne(() => Locations, (location) => location.places)
	place: Locations;

	@typeorm.OneToMany(() => Locations, (location) => location.place)
	places: Locations[];

	@typeorm.ManyToOne(() => Users, (user) => user.locations)
	user: Users;
}
