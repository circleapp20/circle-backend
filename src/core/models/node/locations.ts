import { BaseModel } from 'core/models/node/baseModel';
import { Stores } from 'core/models/node/stores';
import { Users } from 'core/models/node/users';
import * as typeorm from 'typeorm';

@typeorm.Entity()
export class Locations extends BaseModel {
	@typeorm.Column()
	name: string;

	@typeorm.Column('float', { default: 0 })
	latitude: number;

	@typeorm.Column('float', { default: 0 })
	longitude: number;

	@typeorm.Column()
	address: string;

	@typeorm.Column('boolean', { default: false })
	isVerified: boolean;

	@typeorm.ManyToOne(() => Locations, (location) => location.places)
	place: Locations;

	@typeorm.OneToMany(() => Locations, (location) => location.place)
	places: Locations[];

	@typeorm.ManyToOne(() => Users, (user) => user.locations)
	user: Users;

	@typeorm.OneToMany(() => Stores, (store) => store.location)
	stores: Stores[];
}
