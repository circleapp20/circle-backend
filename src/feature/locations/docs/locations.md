# Circle - Locations

Locations in circle all have the same format of name, address, latitude, longitude, isVerified. In view of this, locations must relate to each other, such as country having states, states with cities and cities having streets. To create the relation in the database, a reference column is added called places. This makes it possible to relate locations with other locations based of their id.

In querying the locations to get its, other related locations. Use the SELF join query, to get the related locations or sub locations. Since typeorm is used as our ORM, the query builder format is shown below.

```typescript
export const locationQueryBuilderFormat = (manager: EntityManager) => {
	return manager
		.createQueryBuilder(Locations, 'location')
		.leftJoinAndSelect('location.places', 'states') // get related locations as states
		.leftJoinAndSelect('states.places', 'cities') // get related location as cities for states
		.leftJoinAndSelect('cities.places', 'streets'); // get locations as streets for cities
};
```

To add a new location either by suggestion or by the lead fellow. The details of the location should be added to the database before its related to another location if any.

First of all, locations must be added to the users update endpoint

LOCATIONS
a GET endpoints to return all locations on circle in a JSON format like you already do.
this should include name, nickname and if possible, Google maps coordinates

The return will include:
Main locations (I.e , KNUST and UG for now) with the number of sub locations under them.
The sub locations under each main location and if the sub location has other sub locations, then the number of sub locations should be added and the locations under it.

an endpoint for saving suggested locations by a user. (Name, nickname and description)
