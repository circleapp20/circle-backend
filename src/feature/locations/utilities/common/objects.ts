export const changeLocationPlacesName = (location: any, index = 0) => {
	const { places, ...rest } = location;
	let locationPlaces = places;
	if (locationPlaces.length) {
		// mutation happens here
		locationPlaces = places.map((place: any) => changeLocationPlacesName(place, index + 1));
	}
	const names = ['states', 'cities', 'streets'];
	if (names[index]) rest[names[index]] = locationPlaces;
	return rest;
};
