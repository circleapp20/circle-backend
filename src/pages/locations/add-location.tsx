import { AddLocation } from 'feature/locations/ui/AddLocation';

export default AddLocation;
