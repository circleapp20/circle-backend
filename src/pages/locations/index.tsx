import { Locations } from 'feature/locations/ui/Locations';

export default Locations;
