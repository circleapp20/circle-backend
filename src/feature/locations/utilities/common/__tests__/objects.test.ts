import { locationsList } from 'fixtures/locations';
import { changeLocationPlacesName } from '../objects';

describe('#changeLocationPlacesName', () => {
	test('should rename places', () => {
		const updatedLocations = locationsList.map((location) => {
			const locations = changeLocationPlacesName(location);
			return locations;
		});
		expect(updatedLocations).toMatchSnapshot('updated location places names');
	});
});
