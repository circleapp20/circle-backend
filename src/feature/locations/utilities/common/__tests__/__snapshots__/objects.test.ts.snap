// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#changeLocationPlacesName should rename places: updated location places names 1`] = `
Array [
  Object {
    "address": "",
    "createdAt": "2020-06-20T15:30:58.957Z",
    "id": "84fa52ca-a661-4f73-8202-2fb5ae166bf8",
    "isVerified": false,
    "latitude": 6.973,
    "longitude": 127.398,
    "name": "Ghana",
    "states": Array [
      Object {
        "address": "",
        "cities": Array [
          Object {
            "address": "",
            "createdAt": "2020-06-20T15:32:13.935Z",
            "id": "7a530166-0f77-4528-bbd3-cfa355bed328",
            "isVerified": false,
            "latitude": 6.973,
            "longitude": 127.398,
            "name": "Kumasi",
            "streets": Array [],
            "updatedAt": "2020-06-20T15:32:13.000Z",
          },
        ],
        "createdAt": "2020-06-20T15:31:29.641Z",
        "id": "f039103d-8524-433f-9986-7535c9222464",
        "isVerified": false,
        "latitude": 6.973,
        "longitude": 127.398,
        "name": "Ashanti",
        "updatedAt": "2020-06-20T15:31:29.000Z",
      },
      Object {
        "address": "",
        "cities": Array [],
        "createdAt": "2020-06-20T22:05:52.398Z",
        "id": "81f6240f-4fbd-4f10-856c-f9715ae1da8a",
        "isVerified": false,
        "latitude": 6.973,
        "longitude": 127.398,
        "name": "Greater Accra",
        "updatedAt": "2020-06-20T22:05:52.000Z",
      },
    ],
    "updatedAt": "2020-06-20T15:30:58.957Z",
  },
  Object {
    "address": "",
    "createdAt": "2020-06-20T22:04:25.119Z",
    "id": "0ff59483-a02c-444c-baf1-42b2a443392c",
    "isVerified": false,
    "latitude": 6.973,
    "longitude": 127.398,
    "name": "Togo",
    "states": Array [],
    "updatedAt": "2020-06-20T22:04:25.119Z",
  },
]
`;
